CORE-10427 - Alias Identity Mapper part 2

julia-filipczak opened the pull request:
Adding method to return mapping of alias name and the hardcoded name of the real holding identity and invoking it in the interop processor.

julia-filipczak committed:
adding getRealIdentity method in the interop processor

szymonsztuka commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -40,6 +40,7 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
             //TODO temporary using UnauthenticatedMessage instead of AuthenticatedMessage
             if (unAuthMessage != null && unAuthMessage is UnauthenticatedMessage && unAuthMessage.header.subsystem == SUBSYSTEM) {
                 val header = with(unAuthMessage.header) { CommonHeader(source, destination, null, messageId) }
+                logger.info("The alias ${unAuthMessage.header.destination.x500Name} is mapped to the real holding identity ${getRealHoldingIdentity(unAuthMessage.header.destination).values.firstOrNull()}")
```
Why first and not a second or third?

jmacmahonr3 replied:
@julia-filipczak is the assumption that there will only be 1 matching x500 ?

szymonsztuka commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -99,6 +100,11 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
     } catch (e: NumberFormatException) {
         "${UUID.randomUUID()}"
     }
+    private fun getRealHoldingIdentity(recipientId: net.corda.data.identity.HoldingIdentity): MutableMap<String, String> {
```
The method, name "getRealHoldingIdentity" doesn't do what it says, it doesn't return a HoldingIdentity but a map of something.

szymonsztuka commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -99,6 +100,11 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
     }
+    private fun getRealHoldingIdentity(recipientId: net.corda.data.identity.HoldingIdentity): MutableMap<String, String> {
+        val cache = mutableMapOf<String,String>()
+        cache[recipientId.x500Name.toString()] = "O=Alice,L=London,C=GB"
```
Why this function insert to the cache, should this function get from the cache? Again filling up data to the cache is other concern, e.g. during service startup.

julia-filipczak committed:
addressing pr comments: refactoring getRealIdentity method

szymonsztuka commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -100,6 +107,11 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
         "${UUID.randomUUID()}"
     }
 
+    private fun getRealHoldingIdentity(recipientId: net.corda.data.identity.HoldingIdentity): String? {
```
I think we should return HoldingIdentity object not a string (the same type what is used a parameter).

julia-filipczak committed:
addressing pr comments: changing getRealIdentity return type

szymonsztuka commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -44,7 +45,7 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
                         getRealHoldingIdentity(
                             unAuthMessage.header.destination
-                        )
+                        )?.x500Name.toString()
```
print the whole Holding Identity this is not needed (`?.x500Name.toString())` , think what to print when Identity is null (not found)

julia-filipczak committed:
addressing pr comments: changing logging statement

julia-filipczak committed:
addressing pr comments: refactoring

szymonsztuka commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -100,6 +111,11 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
 
+    private fun getRealHoldingIdentity(recipientId: net.corda.data.identity.HoldingIdentity): HoldingIdentity? {
+        val cache = mutableMapOf<String, HoldingIdentity>()
+        return cache[recipientId.x500Name.toString()]
```
Currently we are returning null, should we return _hardcoded name of the real holding identity_ as written in the PR description?

julia-filipczak replied:
What would this hardcoded name correspond to if we're not actually placing anything in the cache yet?

szymonsztuka replied:
If you are returning null then it's fine, change the PR description, if you want to return something hardcoded then e.g.:
```
private fun getRealHoldingIdentity(recipientId: net.corda.data.identity.HoldingIdentity): HoldingIdentity? {
        return HoldingIdentity(MemberX500Name.parse("CN=Alice, O=Alice Corp, L=LDN, C=GB"), "Gold")
}
```
or 
```
private fun getRealHoldingIdentity(recipientId: net.corda.data.identity.HoldingIdentity): HoldingIdentity? {
        val cache = mutableMapOf<String, HoldingIdentity>()
        // As the cache is empty returning hardcoded Alice from Gold for now
        return cache[recipientId.x500Name.toString()] ? HoldingIdentity(MemberX500Name.parse("CN=Alice, O=Alice Corp, L=LDN, C=GB"), "Gold")
}
```
or
```
private fun getRealHoldingIdentity(recipientId: net.corda.data.identity.HoldingIdentity): HoldingIdentity? {
        val cache = mutableMapOf<String, HoldingIdentity>()
        // As the cache is empty returning the same Identity with hardoced group "Gold" 
        return cache[recipientId.x500Name.toString()] ? HoldingIdentity(recipientId.x500Name, "Gold")
}
```

jmacmahonr3 replied:
Create a method that gets called on processor start up called loadAliasInfo and create the hardcoded map there. 
Then in your getRealIdentity method just call that map.

julia-filipczak committed:
addressing pr comments: adding hardcoded value to be returned when the cache is null

julia-filipczak committed:
adding the functionality to read from the hosted identities and member list topics to populate the cache and return the mapping in the interop processor

yarivus commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/service/impl/InteropAliasTranslatorImpl.kt:
```diff
@@ -0,0 +1,102 @@
+    configuration: SmartConfig
+) : InteropAliasTranslator {
+    companion object {
+        private const val GROUP_NAME = "interop_stub_hosting_map"
```
Is this the right group name ?
Shouldn't the name contain alias & translator ?

yarivus commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/service/impl/InteropAliasTranslatorImpl.kt:
```diff
@@ -0,0 +1,102 @@
+import java.util.concurrent.CompletableFuture
+import java.util.concurrent.ConcurrentHashMap
+
+class InteropAliasTranslatorImpl(
```
I'm assuming you're planning a unit-test for this class

jmacmahonr3 replied:
Please add a comment to this class explaining why we need it also please link to the Ticket from milestone 4 where the db worker will write to memberInfo topic which will allow us to delete this class.

yarivus reviewed (commented):
Checkpointing review

jmacmahonr3 commented on components/interop/interop-service/build.gradle:
```diff
@@ -14,6 +14,9 @@ dependencies {
     implementation project(":libs:lifecycle:lifecycle")
     implementation project(":libs:messaging:messaging")
     implementation project(":libs:flows:session-manager")
+    implementation project(":components:link-manager")
```
What are we dependent on from link manager ?

jmacmahonr3 commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -42,10 +48,13 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
             //TODO temporary using UnauthenticatedMessage instead of AuthenticatedMessage
             if (unAuthMessage != null && unAuthMessage is UnauthenticatedMessage && unAuthMessage.header.subsystem == SUBSYSTEM) {
                 val header = with(unAuthMessage.header) { CommonHeader(source, destination, null, messageId) }
+                val groupReader = membershipGroupReaderProvider.getGroupReader(unAuthMessage.header.destination.toCorda())
```
these three lines can be extracted to a method

julia-filipczak committed:
Merge branch 'interop/5.0' into juliaf/CORE-10427-3

julia-filipczak committed:
refactoring to use osgi annotation in the translator class

jmacmahonr3 commented:
@julia-filipczak is interop/5.0 the correct branch ? I thought that that was the milestone 1 branch ?

szymonsztuka commented:
> @julia-filipczak is interop/5.0 the correct branch ? I thought that that was the milestone 1 branch ?

This is the only branch having Interop code

jmacmahonr3 commented:
> > @julia-filipczak is interop/5.0 the correct branch ? I thought that that was the milestone 1 branch ?
> 
> This is the only branch having Interop code

Didn't Saurabh cut another branch? 
We should have one branch per milestone going forward. This will enable Dave M to test without the code changing under his feet.

julia-filipczak committed:
further refactoring to resolve osgi errors

julia-filipczak committed:
fixing gradle detekt error

SaurabhR3 commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -30,8 +44,10 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
+        InteropAliasTranslatorImpl(lifecycleCoordinatorFactory, subscriptionFactory, config)
 
-    override fun onNext(
+        override fun onNext(
```
Extra space

SaurabhR3 commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -44,6 +60,13 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
+                    getRealHoldingIdentityFromAliasMapping(unAuthMessage)
+                )
+            }"
+        )
```
You can remove extra line here

SaurabhR3 commented on components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropProcessor.kt:
```diff
@@ -100,6 +123,12 @@ class InteropProcessor(cordaAvroSerializationFactory: CordaAvroSerializationFact
+        val memberInfo = groupReader.lookup(unAuthMessage.header.destination.toCorda().x500Name)
+        return memberInfo?.memberProvidedContext?.get(MemberInfoExtension.INTEROP_ALIAS_MAPPING)
+    }
+
```
Better to just accept `destination: String` as parameter rather than complete object `unAuthMessage: UnauthenticatedMessage`

julia-filipczak committed:
further refactoring

julia-filipczak committed:
reverting refactoring as it caused detekt issue

julia-filipczak committed:
Merge branch 'interop/5.0' into juliaf/CORE-10427-3

julia-filipczak committed:
refactoring

yarivus reviewed (commented):
no comments - checkpointing review

julia-filipczak committed:
fix gradle detekt

SaurabhR3 committed:
CORE:10427 : 1. Changes for Membership group read impl service
2. BND dependency to make integration test happy
3. Interop Alias mapping processor
4. Removed domino tile processor
5. Changed integration test data

yarivus reviewed (commented):
checkpointing review - looks good

SaurabhR3 committed:
CORE:10427 : Fixed test

julia-filipczak committed:
minor cleanup and updating cordaApiVersion to include the changes made in corda-api

SaurabhR3 committed:
CORE:10427 : Removed not required logs

julia-filipczak committed:
fixing version number

julia-filipczak committed:
Merge remote-tracking branch 'origin/juliaf/CORE-10427-3' into juliaf/CORE-10427-3

yarivus reviewed (approved):
LGTM

julia-filipczak committed:
possible fix for jenkins build failure during cluster deployment

yarivus reviewed (approved):
LGTM

julia-filipczak committed:
Merge remote-tracking branch 'origin/release/interop/rudiments' into juliaf/CORE-10427-3

# Conflicts:
#	components/interop/interop-service/src/main/kotlin/net/corda/interop/InteropService.kt
#	components/interop/interop-service/test.bndrun
#	gradle.properties

jmacmahonr3 reviewed (approved):
LGTM

julia-filipczak committed:
update

yarivus reviewed (approved):
LGTM

julia-filipczak committed:
Merge remote-tracking branch 'origin/release/interop/rudiments' into juliaf/CORE-10427-3

# Conflicts:
#	gradle.properties

julia-filipczak committed:
updating corda-api version

yarivus reviewed (approved):
LGTM
